From c7aa527f01b6df9596b0b9649536165b8a4c8933 Mon Sep 17 00:00:00 2001
From: Xiaoyu Huang <quic_xiahua@quicinc.com>
Date: Thu, 21 Nov 2024 12:22:06 +0000
Subject: [PATCH] Weston: support SDM backend on weston-13.0.1

1, Customized modifications to support the detached sdm-backend,
including exporting hearder files/static libraries
2, Add support for gbm-buffer-backend and screen-capture
3, Add Weston/GFX KPI markers

Upstream-Status: Inappropriate [qti specific]

Change-Id: I2a87cb5ffaff3abcd701a975066265ed2299a168
Signed-off-by: Xiaoyu Huang <quic_xiahua@quicinc.com>
---
 compositor/main.c                            |   7 +
 include/libweston/libweston.h                |  27 +
 libweston/backend-drm/libbacklight.c         |  18 +-
 libweston/backend-drm/meson.build            |   9 +-
 libweston/compositor.c                       |  61 ++
 libweston/libinput-seat.c                    |  15 +-
 libweston/libweston-internal.h               |  14 +
 libweston/log.c                              |  17 +
 libweston/meson.build                        |  26 +-
 libweston/renderer-gl/gl-renderer-internal.h |   9 +
 libweston/renderer-gl/gl-renderer.c          | 672 ++++++++++++++++++-
 libweston/renderer-gl/meson.build            |   9 +-
 meson.build                                  |   5 +
 shared/meson.build                           |   8 +
 14 files changed, 875 insertions(+), 22 deletions(-)

diff --git a/compositor/main.c b/compositor/main.c
index 0e3d375..4cf54d7 100644
--- a/compositor/main.c
+++ b/compositor/main.c
@@ -26,6 +26,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #include "config.h"
@@ -4083,6 +4088,8 @@ wet_main(int argc, char *argv[], const struct weston_testsuite_data *test_data)
 		{ WESTON_OPTION_STRING, "flight-rec-scopes", 'f', &flight_rec_scopes },
 	};
 
+	weston_place_marker("W - weston main begin");
+
 	wl_list_init(&wet.layoutput_list);
 	wl_list_init(&wet.backend_list);
 
diff --git a/include/libweston/libweston.h b/include/libweston/libweston.h
index 077c16b..accca79 100644
--- a/include/libweston/libweston.h
+++ b/include/libweston/libweston.h
@@ -23,6 +23,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #ifndef _WAYLAND_SYSTEM_COMPOSITOR_H_
@@ -73,6 +78,11 @@ struct weston_point2d_device_normalized {
 	double y;
 };
 
+struct surface_color {
+	bool is_pended;
+	float red, blue, green, alpha;
+};
+
 struct weston_surface;
 struct weston_buffer;
 struct shell_surface;
@@ -89,6 +99,7 @@ struct weston_color_transform;
 struct pixel_format_info;
 struct weston_output_capture_info;
 struct weston_tearing_control;
+struct gbm_buffer;
 
 enum weston_keyboard_modifier {
 	MODIFIER_CTRL = (1 << 0),
@@ -565,8 +576,12 @@ struct weston_output {
 			  uint16_t *g,
 			  uint16_t *b);
 
+	void (*enable_ppm)(struct weston_output *output, int32_t enable);
+	void (*set_ppm)(struct weston_output *output, int32_t ppm);
 	bool enabled; /**< is in the output_list, not pending list */
 	int scale;
+	/* Indicate current repainting need gpu do compositon. */
+	bool need_gpu_composition;
 
 	struct weston_color_profile *color_profile;
 	bool from_blend_to_output_by_backend;
@@ -1531,6 +1546,7 @@ struct weston_buffer {
 	enum {
 		WESTON_BUFFER_SHM,
 		WESTON_BUFFER_DMABUF,
+		WESTON_BUFFER_GBMBUF,
 		WESTON_BUFFER_RENDERER_OPAQUE,
 		WESTON_BUFFER_SOLID,
 	} type;
@@ -1538,6 +1554,7 @@ struct weston_buffer {
 	union {
 		struct wl_shm_buffer *shm_buffer;
 		void *dmabuf;
+		void *gbmbuf;
 		void *legacy_buffer;
 		struct weston_solid_buffer_values solid;
 	};
@@ -1738,6 +1755,12 @@ struct weston_view {
 
 	bool is_mapped;
 	struct weston_log_pacer subsurface_parent_log_pacer;
+
+	/* Indicate if this view is used for screen capture. */
+	bool is_capture_view;
+
+	/* Indicate if this view is completely covered by above opaque regions. */
+	bool is_completely_covered;
 };
 
 enum weston_surface_status {
@@ -1934,6 +1957,7 @@ struct weston_surface {
 
 	/* An list of per seat pointer constraints. */
 	struct wl_list pointer_constraints;
+	struct surface_color surf_color;
 
 	/* zwp_surface_synchronization_v1 resource for this surface */
 	struct wl_resource *synchronization_resource;
@@ -2384,6 +2408,9 @@ weston_log_paced(struct weston_log_pacer *pacer, unsigned int max_burst,
 		 unsigned int reset_ms, const char *fmt, ...)
 	__attribute__ ((format (printf, 4, 5)));
 
+void
+weston_place_marker(const char *name);
+
 enum weston_screenshooter_outcome {
 	WESTON_SCREENSHOOTER_SUCCESS,
 	WESTON_SCREENSHOOTER_NO_MEMORY,
diff --git a/libweston/backend-drm/libbacklight.c b/libweston/backend-drm/libbacklight.c
index 74690fa..4d4def7 100644
--- a/libweston/backend-drm/libbacklight.c
+++ b/libweston/backend-drm/libbacklight.c
@@ -27,6 +27,11 @@
  * Authors:
  *    Matthew Garrett <mjg@redhat.com>
  *    Tiago Vignatti <vignatti@freedesktop.org>
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #include "config.h"
@@ -44,6 +49,7 @@
 #include <libgen.h>
 #include <string.h>
 #include <errno.h>
+#include <wayland-util.h>
 
 #include "shared/string-helpers.h"
 
@@ -87,22 +93,22 @@ out:
 	return ret;
 }
 
-long backlight_get_brightness(struct backlight *backlight)
+WL_EXPORT long backlight_get_brightness(struct backlight *backlight)
 {
 	return backlight_get(backlight, "brightness");
 }
 
-long backlight_get_max_brightness(struct backlight *backlight)
+WL_EXPORT long backlight_get_max_brightness(struct backlight *backlight)
 {
 	return backlight_get(backlight, "max_brightness");
 }
 
-long backlight_get_actual_brightness(struct backlight *backlight)
+WL_EXPORT long backlight_get_actual_brightness(struct backlight *backlight)
 {
 	return backlight_get(backlight, "actual_brightness");
 }
 
-long backlight_set_brightness(struct backlight *backlight, long brightness)
+WL_EXPORT long backlight_set_brightness(struct backlight *backlight, long brightness)
 {
 	char *path;
 	char *buffer = NULL;
@@ -147,7 +153,7 @@ out:
 	return ret;
 }
 
-void backlight_destroy(struct backlight *backlight)
+WL_EXPORT void backlight_destroy(struct backlight *backlight)
 {
 	if (!backlight)
 		return;
@@ -158,7 +164,7 @@ void backlight_destroy(struct backlight *backlight)
 	free(backlight);
 }
 
-struct backlight *backlight_init(struct udev_device *drm_device,
+WL_EXPORT struct backlight *backlight_init(struct udev_device *drm_device,
 				 uint32_t connector_type)
 {
 	const char *syspath = NULL;
diff --git a/libweston/backend-drm/meson.build b/libweston/backend-drm/meson.build
index e048789..d90c2cf 100644
--- a/libweston/backend-drm/meson.build
+++ b/libweston/backend-drm/meson.build
@@ -13,7 +13,7 @@ dep_libdisplay_info = dependency(
 )
 config_h.set10('HAVE_LIBDISPLAY_INFO', dep_libdisplay_info.found())
 
-lib_backlight = static_library(
+lib_backlight = shared_library(
 	'backlight',
 	'libbacklight.c',
 	dependencies: [
@@ -21,7 +21,7 @@ lib_backlight = static_library(
 		dependency('libudev')
 	],
 	include_directories: common_inc,
-	install: false
+	install: true
 )
 dep_backlight = declare_dependency(
 	link_with: lib_backlight,
@@ -101,3 +101,8 @@ plugin_drm = shared_library(
 env_modmap += 'drm-backend.so=@0@;'.format(plugin_drm.full_path())
 
 install_headers(backend_drm_h, subdir: dir_include_libweston_install)
+
+install_headers(
+	'libbacklight.h',
+	subdir: dir_include_libweston_private_install
+)
diff --git a/libweston/compositor.c b/libweston/compositor.c
index 41459eb..c047a2a 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -25,6 +25,12 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
+/*
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
 
 #include "config.h"
 
@@ -60,6 +66,7 @@
 #include <libweston/weston-log.h>
 #include "linux-dmabuf.h"
 #include "linux-dmabuf-unstable-v1-server-protocol.h"
+#include "gbm-buffer-backend.h"
 #include "viewporter-server-protocol.h"
 #include "presentation-time-server-protocol.h"
 #include "xdg-output-unstable-v1-server-protocol.h"
@@ -95,6 +102,8 @@
 
 #define DEFAULT_REPAINT_WINDOW 7 /* milliseconds */
 
+struct gbm_buffer_backend_c_interface *gbm_buffer_backend;
+
 static void
 weston_output_transform_scale_init(struct weston_output *output,
 				   uint32_t transform, uint32_t scale);
@@ -600,6 +609,7 @@ weston_view_create_internal(struct weston_surface *surface)
 	pixman_region32_init(&view->geometry.scissor);
 	pixman_region32_init(&view->transform.boundingbox);
 	view->transform.dirty = 1;
+	view->is_completely_covered = false;
 	weston_view_update_transform(view);
 	pixman_region32_copy(&view->visible, &view->transform.boundingbox);
 
@@ -2591,6 +2601,7 @@ weston_buffer_from_resource(struct weston_compositor *ec,
 	struct weston_buffer *buffer;
 	struct wl_shm_buffer *shm;
 	struct linux_dmabuf_buffer *dmabuf;
+	struct gbm_buffer *gbmbuf;
 	struct wl_listener *listener;
 	struct weston_solid_buffer_values *solid;
 
@@ -2640,6 +2651,25 @@ weston_buffer_from_resource(struct weston_compositor *ec,
 			buffer->buffer_origin = ORIGIN_BOTTOM_LEFT;
 		else
 			buffer->buffer_origin = ORIGIN_TOP_LEFT;
+	} else if (gbm_buffer_backend && (gbmbuf = gbm_buffer_backend->buffer_get(buffer->resource))) {
+		buffer->type = WESTON_BUFFER_GBMBUF;
+		buffer->gbmbuf = gbmbuf;
+		buffer->width = gbmbuf->width;
+		buffer->height = gbmbuf->height;
+		buffer->pixel_format =
+			pixel_format_get_info(gbmbuf->format);
+		/* gbmbuf import should assure we don't create a buffer with an
+		 * unknown format */
+		assert(buffer->pixel_format && !buffer->pixel_format->hide_from_clients);
+		/*
+		* To be aligned with dmabuf buffers, since gbmbuf buffers also
+		* have the origin at top-left, invert the Y_INVERT flag to get
+		* the image right.
+		*/
+		if (gbmbuf->flags & ZWP_LINUX_BUFFER_PARAMS_V1_FLAGS_Y_INVERT)
+			buffer->buffer_origin = ORIGIN_BOTTOM_LEFT;
+		else
+			buffer->buffer_origin = ORIGIN_TOP_LEFT;
 	} else if ((solid = single_pixel_buffer_get(buffer->resource))) {
 		buffer->type = WESTON_BUFFER_SOLID;
 		buffer->solid = *solid;
@@ -9391,6 +9421,31 @@ weston_compositor_dmabuf_can_scanout(struct weston_compositor *compositor,
 	return true;
 }
 
+/** Import gbmbuf buffer into current renderer
+ *
+ * \param compositor
+ * \param buffer the gbmbuf buffer to import
+ * \return true on usable buffers, false otherwise
+ *
+ * This function tests that the gbm_buffer is usable
+ * for the current renderer. Returns false on unusable buffers. Usually
+ * usability is tested by importing the gbmbuf for composition.
+ *
+ * This hook is also used for detecting if the renderer supports
+ * gbmbuf at all. If the renderer hook is NULL, dmabufs are not
+ * supported.
+ * */
+WL_EXPORT bool
+weston_compositor_import_gbm_buffer(struct weston_compositor *compositor,
+				struct gbm_buffer *buffer)
+{
+	struct weston_renderer *renderer;
+	renderer = compositor->renderer;
+	if (renderer->import_gbm_buffer == NULL)
+		return false;
+	return renderer->import_gbm_buffer(compositor, buffer);
+}
+
 WL_EXPORT void
 weston_version(int *major, int *minor, int *micro)
 {
@@ -9722,6 +9777,12 @@ weston_compositor_init_renderer(struct weston_compositor *compositor,
 
 		compositor->renderer->gl = gl_renderer;
 		weston_log("Using GL renderer\n");
+
+		gbm_buffer_backend = weston_load_module("gbm-buffer-backend.so",
+						"gbm_buffer_backend_c_interface",
+						LIBWESTON_MODULEDIR);
+		assert(gbm_buffer_backend);
+
 		break;
 	case WESTON_RENDERER_PIXMAN:
 		ret = pixman_renderer_init(compositor);
diff --git a/libweston/libinput-seat.c b/libweston/libinput-seat.c
index 5ecb974..c9d4130 100644
--- a/libweston/libinput-seat.c
+++ b/libweston/libinput-seat.c
@@ -22,6 +22,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #include "config.h"
@@ -173,7 +178,7 @@ udev_seat_remove_devices(struct udev_seat *seat)
 	}
 }
 
-void
+WL_EXPORT void
 udev_input_disable(struct udev_input *input)
 {
 	if (input->suspended)
@@ -270,7 +275,7 @@ const struct libinput_interface libinput_interface = {
 	close_restricted,
 };
 
-int
+WL_EXPORT int
 udev_input_enable(struct udev_input *input)
 {
 	struct wl_event_loop *loop;
@@ -333,7 +338,7 @@ libinput_log_func(struct libinput *libinput,
 	weston_vlog(format, args);
 }
 
-int
+WL_EXPORT int
 udev_input_init(struct udev_input *input, struct weston_compositor *c,
 		struct udev *udev, const char *seat_id,
 		udev_configure_device_t configure_device)
@@ -378,7 +383,7 @@ udev_input_init(struct udev_input *input, struct weston_compositor *c,
 	return udev_input_enable(input);
 }
 
-void
+WL_EXPORT void
 udev_input_destroy(struct udev_input *input)
 {
 	struct udev_seat *seat, *next;
@@ -489,7 +494,7 @@ udev_seat_destroy(struct udev_seat *seat)
 	free(seat);
 }
 
-struct udev_seat *
+WL_EXPORT struct udev_seat *
 udev_seat_get_named(struct udev_input *input, const char *seat_name)
 {
 	struct udev_seat *seat;
diff --git a/libweston/libweston-internal.h b/libweston/libweston-internal.h
index 0cb98dc..44b173f 100644
--- a/libweston/libweston-internal.h
+++ b/libweston/libweston-internal.h
@@ -23,6 +23,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #ifndef LIBWESTON_INTERNAL_H
@@ -70,6 +75,9 @@ struct weston_renderer {
 	void (*repaint_output)(struct weston_output *output,
 			       pixman_region32_t *output_damage,
 			       struct weston_renderbuffer *renderbuffer);
+	void (*capture_screen)(struct weston_output *output,
+			       struct weston_buffer *buffer,
+			       pixman_region32_t *output_damage);
 
 	/** See weston_renderer_resize_output()
 	 *
@@ -91,6 +99,9 @@ struct weston_renderer {
 				    int src_x, int src_y,
 				    int width, int height);
 
+	/** See weston_compositor_import_gbm_buffer() */
+	bool (*import_gbm_buffer)(struct weston_compositor *ec,
+			        struct gbm_buffer *buffer);
 	/** See weston_compositor_import_dmabuf() */
 	bool (*import_dmabuf)(struct weston_compositor *ec,
 			      struct linux_dmabuf_buffer *buffer);
@@ -175,6 +186,9 @@ weston_compositor_import_dmabuf(struct weston_compositor *compositor,
 bool
 weston_compositor_dmabuf_can_scanout(struct weston_compositor *compositor,
 					struct linux_dmabuf_buffer *buffer);
+bool
+weston_compositor_import_gbm_buffer(struct weston_compositor *compositor,
+				struct gbm_buffer *buffer);
 void
 weston_compositor_offscreen(struct weston_compositor *compositor);
 
diff --git a/libweston/log.c b/libweston/log.c
index a949da4..fd9d6e9 100644
--- a/libweston/log.c
+++ b/libweston/log.c
@@ -21,6 +21,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #include "config.h"
@@ -31,6 +36,7 @@
 #include <string.h>
 #include <sys/time.h>
 #include <time.h>
+#include <fcntl.h>
 
 #include <wayland-util.h>
 
@@ -253,3 +259,14 @@ weston_log_continue(const char *fmt, ...)
 
 	return l;
 }
+
+WL_EXPORT void
+weston_place_marker(const char *name)
+{
+	int fd = open("/sys/kernel/boot_kpi/kpi_values", O_WRONLY);
+
+	if (fd > 0) {
+		write(fd, name, strlen(name));
+		close(fd);
+	}
+}
diff --git a/libweston/meson.build b/libweston/meson.build
index a83a52d..3b841b6 100644
--- a/libweston/meson.build
+++ b/libweston/meson.build
@@ -5,6 +5,7 @@ deps_libweston = [
 	dep_libdl,
 	dep_libdrm,
 	dep_xkbcommon,
+	dep_gbm,
 	dep_matrix_c
 ]
 srcs_libweston = [
@@ -171,19 +172,20 @@ if get_option('backend-drm')
 	deps_session_helper = [
 		dep_libseat,
 		dep_libweston_private_h,
+		dep_libweston_public,
 	]
 
-	lib_session_helper = static_library(
+	lib_session_helper = shared_library(
 		'session-helper',
 		srcs_session_helper,
 		include_directories: common_inc,
 		dependencies: deps_session_helper,
-		install: false
+		install: true
 	)
 	dep_session_helper = declare_dependency(link_with: lib_session_helper)
 endif
 
-lib_libinput_backend = static_library(
+lib_libinput_backend = shared_library(
 	'libinput-backend',
 	[
 		'libinput-device.c',
@@ -193,10 +195,11 @@ lib_libinput_backend = static_library(
 	dependencies: [
 		dep_libweston_private,
 		dep_libinput,
-		dependency('libudev', version: '>= 136')
+		dependency('libudev', version: '>= 136'),
+		dep_session_helper
 	],
 	include_directories: common_inc,
-	install: false
+	install: true
 )
 dep_libinput_backend = declare_dependency(
 	link_with: lib_libinput_backend,
@@ -225,6 +228,19 @@ dep_lib_gl_borders = declare_dependency(
 	dependencies: dep_lib_cairo_shared
 )
 
+install_headers(
+	'backend.h',
+	'libinput-seat.h',
+	'libweston-internal.h',
+	'linux-dmabuf.h',
+	'linux-explicit-synchronization.h',
+	'color.h',
+	'launcher-util.h',
+	'pixman-renderer.h',
+	'pixel-formats.h',
+	subdir: dir_include_libweston_private_install
+)
+
 subdir('color-lcms')
 subdir('renderer-gl')
 subdir('backend-drm')
diff --git a/libweston/renderer-gl/gl-renderer-internal.h b/libweston/renderer-gl/gl-renderer-internal.h
index 5032035..5840254 100644
--- a/libweston/renderer-gl/gl-renderer-internal.h
+++ b/libweston/renderer-gl/gl-renderer-internal.h
@@ -24,6 +24,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries. 
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #ifndef GL_RENDERER_INTERNAL_H
@@ -141,6 +146,9 @@ struct gl_renderer {
 	EGLContext egl_context;
 	EGLConfig egl_config;
 
+	/* gbm device handle*/
+	struct gbm_device *gbm_hdle;
+
 	uint32_t gl_version;
 
 	struct wl_array vertices;
@@ -180,6 +188,7 @@ struct gl_renderer {
 	bool has_dmabuf_import;
 	struct wl_list dmabuf_images;
 	struct wl_list dmabuf_formats;
+	struct wl_list gbmbuf_images;
 
 	bool has_texture_type_2_10_10_10_rev;
 	bool has_gl_texture_rg;
diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index faf3bdf..7825356 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -23,6 +23,11 @@
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
+ *
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #include "config.h"
@@ -63,6 +68,9 @@
 #include "shared/weston-drm-fourcc.h"
 #include "shared/weston-egl-ext.h"
 #include "shared/xalloc.h"
+#include "gbm_priv.h"
+#include "gbm-buffer-backend.h"
+#include "screen-capture.h"
 
 #define BUFFER_DAMAGE_COUNT 2
 
@@ -206,6 +214,9 @@ struct timeline_render_point {
 	struct wl_event_source *event_source;
 };
 
+struct gbm_buffer_backend_c_interface *gbm_buffer_backend;
+struct screen_capture_c_interface *screen_capture_backend;
+
 static uint32_t
 gr_gl_version(uint16_t major, uint16_t minor)
 {
@@ -1360,18 +1371,169 @@ out:
 	pixman_region32_fini(&repaint);
 }
 
+/* clear_paint_node: It composes onto GPU composed layer a clear transparent
+					 bounding box of the size same as that of overlay layer
+					 to allow overlay plane to be visible through the GPU
+					 composed layer.
+   Inputs:
+					 struct weston_paint_node *pnode	  : weston paint node
+					 pixman_region32_t	*damage
+   Output:
+					 GPU composed layer pnode->surface will have a clear
+					 transparent bounding box with alpha = 0 and all
+					 3 color channels also equal to 0.
+
+ * For more details of changing from weston_view to weston_paint_node, refers to
+   https://gitlab.freedesktop.org/wayland/weston/-/commit/da0f7ea4a7539e3d4fcd21d87303412a060aa6c8
+*/
+static void
+clear_paint_node(struct weston_paint_node *pnode,
+	   pixman_region32_t *damage) /* in global coordinates */
+{
+	struct gl_renderer *gr = get_renderer(pnode->surface->compositor);
+	struct gl_surface_state *gs = get_surface_state(pnode->surface);
+	struct weston_buffer *buffer = gs->buffer_ref.buffer;
+	struct gl_buffer_state *gb;
+	/* repaint bounding region in global coordinates: */
+	pixman_region32_t repaint;
+	/* non-opaque region in surface coordinates: */
+	pixman_region32_t surface_blend;
+	GLint filter;
+	GLfloat color[4];
+	enum gl_shader_texture_variant shader_variant;
+	int pitch, height;
+	int i;
+
+	assert(gs->buffer);
+	gb = gs->buffer;
+
+	if (gb->shader_variant == SHADER_VARIANT_NONE && !buffer->direct_display)
+		return;
+
+	/* gl_renderer_surface_set_color() called below will reset the surface_state,
+	 * so we save the original ones here. */
+	shader_variant = gb->shader_variant;
+	pitch = gb->pitch;
+	for (i = 0; i < 4; i++)
+		color[i] = gb->color[i];
+
+	pixman_region32_init(&repaint);
+	/* clear boundingbox region of the overlay surface */
+	pixman_region32_copy(&repaint, &pnode->view->transform.boundingbox);
+
+	if (!pixman_region32_not_empty(&repaint))
+		goto out_clear_paint_node;
+
+	glDisable(GL_BLEND);
+
+	gb->color[0] = 0.0f;
+	gb->color[1] = 0.0f;
+	gb->color[2] = 0.0f;
+	gb->color[3] = 0.0f;
+
+	gb->shader_variant = SHADER_VARIANT_SOLID;
+
+	if (pnode->view->transform.enabled ||
+	    pnode->output->current_scale != pnode->surface->buffer_viewport.buffer.scale)
+		filter = GL_LINEAR;
+	else
+		filter = GL_NEAREST;
+
+	struct gl_shader_config sconf;
+	if (!gl_shader_config_init_for_paint_node(&sconf, pnode, filter))
+		goto out_clear_paint_node;
+	maybe_censor_override(&sconf, pnode);
+
+	/* Reset sconf.input_tex since we have chosen SOLID shader */
+	for (i = 0; i < gb->num_textures; i++)
+		sconf.input_tex[i] = 0;
+
+	pixman_region32_init_rect(&surface_blend, 0, 0,
+				pnode->surface->width, pnode->surface->height);
+	repaint_region(gr, pnode, &repaint, &surface_blend, &sconf);
+
+	pixman_region32_fini(&surface_blend);
+
+	/* restore gs setting */
+	gb->shader_variant = shader_variant;
+	gb->pitch = pitch;
+	for (i = 0; i < 4; i++)
+		gb->color[i] = color[i];
+
+out_clear_paint_node:
+	pixman_region32_fini(&repaint);
+}
+
+static void
+draw_paint_node_overlay(struct weston_paint_node *pnode, pixman_region32_t *damage,
+				bool have_primary_view)
+{
+	pixman_region32_t r;
+	bool is_yuv;
+	struct gbm_buffer *gbmbuf;
+	struct linux_dmabuf_buffer *dmabuf;
+	struct weston_compositor *ec = pnode->surface->compositor;
+
+	/* The invisible layer could be marked as SDE by strategy, which means both gpu
+	 * and MDP could ignore this kind of view, no hardware pipe for it. But the buffer
+	 * could be NULL in this case, add this judgement to avoid crash.
+	 */
+	if (pnode->surface->buffer_ref.buffer == NULL) {
+	/* TODO: check why it's happen since invisible layers should be handled
+	 * by view->is_completely_covered.
+	 */
+		return;
+	}
+	/* this view is composed directly by overlay */
+	/* compute whether this view has blending */
+	pixman_region32_init_rect(&r, 0, 0, pnode->surface->width,
+			pnode->surface->height);
+	pixman_region32_subtract(&r, &r, &pnode->surface->opaque);
+
+	/* get yuv statement from gbm or dma buffer info */
+	is_yuv = false;
+	struct weston_buffer *buffer = pnode->surface->buffer_ref.buffer;
+	if (gbmbuf = gbm_buffer_backend->buffer_get(buffer->resource))
+		is_yuv = gbm_buffer_backend->is_yuv_format(gbmbuf->format);
+	else if (dmabuf = linux_dmabuf_buffer_get(ec, buffer->resource))
+		is_yuv = gbm_buffer_backend->is_yuv_format(dmabuf->attributes.format);
+
+	/* only can clear view by meeting all the three conditions:
+	 * 1, have views on primary plane. If none of views composed by gpu,
+	 * don't clear framebuffer becuase it had been cleared before.
+	 * 2, the whole surface region is opaque or it's yuv region.
+	 * 3, global alpha value is 1.
+	 */
+	if (have_primary_view && (!pixman_region32_not_empty(&r) || is_yuv) &&
+			(pnode->view->alpha == 1)) {
+	/* clear framebuffer with transparent pixels where this layer would be*/
+		clear_paint_node(pnode, damage);
+	}
+	pixman_region32_fini(&r);
+}
+
 static void
 repaint_views(struct weston_output *output, pixman_region32_t *damage)
 {
 	struct weston_paint_node *pnode;
+	bool have_primary_view = false;
 
 	glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
 	glEnableVertexAttribArray(0);
 
 	wl_list_for_each_reverse(pnode, &output->paint_node_z_order_list,
 				 z_order_link) {
-		if (pnode->plane == &output->primary_plane)
+		/* Skip screen capture buffer during GPU composition */
+		if (screen_capture_backend->is_screen_capture_view(pnode->view))
+			continue;
+		if (pnode->view->is_completely_covered)
+			continue;
+		if (pnode->plane == &output->primary_plane) {
+			have_primary_view = true;
 			draw_paint_node(pnode, damage);
+		} else {
+			draw_paint_node_overlay(pnode, damage, have_primary_view);
+		}
 	}
 
 	glDisableVertexAttribArray(0);
@@ -2060,6 +2222,145 @@ gl_renderer_repaint_output(struct weston_output *output,
 	gl_renderer_garbage_collect_programs(gr);
 }
 
+/*
+ * Capture screen content into a specified buffer by using FBO. This buffer
+ * must be gbm buffer which is created by GBM buffer protocol, otherwise, egl
+ * image must be created here.
+ */
+static void
+gl_renderer_capture_screen(struct weston_output *output,
+                              struct weston_buffer *buffer,
+                              pixman_region32_t *output_damage)
+{
+	struct gl_output_state *go = get_output_state(output);
+	struct weston_compositor *compositor = output->compositor;
+	GLuint framebuffer, texture;
+	GLenum status;
+	struct gbm_buffer *gbm_buf = NULL;
+	EGLImageKHR cap_buf_image;
+	struct gl_buffer_state *gb;
+	struct weston_view *view;
+	struct weston_paint_node *pnode;
+	const int32_t area_inv_y =
+		go->fb_size.height - go->area.y - go->area.height;
+	struct gl_renderbuffer *rb;
+
+	if (!buffer) {
+		weston_log("Error! buffer is NULL.\n");
+		return;
+	}
+
+	if (wl_shm_buffer_get(buffer->resource) ||
+		linux_dmabuf_buffer_get(compositor, buffer->resource)) {
+		weston_log("Error! buffer is not supported by screen capture.\n");
+		return;
+	}
+
+	if (gbm_buf = gbm_buffer_backend->buffer_get(buffer->resource)) {
+		gb = buffer->renderer_private;
+		cap_buf_image = gb->images[0];
+	} else if (gbm_buf = wl_resource_get_user_data(buffer->resource)) {
+		/* TODO: Create egl image */
+		weston_log("Error! no egl image is bound.\n");
+		return;
+	}
+
+	if (cap_buf_image == EGL_NO_IMAGE_KHR) {
+		weston_log("Error! no egl image is bound.\n");
+		return;
+	}
+
+	/* Accumulate damage in all renderbuffers */
+	wl_list_for_each(rb, &go->renderbuffer_list, link) {
+		pixman_region32_union(&rb->base.damage,
+				      &rb->base.damage,
+				      output_damage);
+		rb->border_damage |= go->border_status;
+	}
+
+	rb = output_get_dummy_renderbuffer(output);
+
+
+	/* Clear the used_in_output_repaint flag, so that we can properly track
+	 * which surfaces were used in this output repaint. */
+	wl_list_for_each_reverse(pnode, &output->paint_node_z_order_list,
+				 z_order_link) {
+		if (pnode->plane == &output->primary_plane) {
+			struct gl_surface_state *gs =
+				get_surface_state(pnode->view->surface);
+			gs->used_in_output_repaint = false;
+		}
+	}
+
+	/* If using shadow, redirect all drawing to it first. */
+	if (shadow_exists(go)) {
+		glBindFramebuffer(GL_FRAMEBUFFER, go->shadow.fbo);
+		glViewport(0, 0, go->area.width, go->area.height);
+	} else {
+		glBindFramebuffer(GL_FRAMEBUFFER, rb->fbo);
+		glViewport(go->area.x, area_inv_y,
+			   go->area.width, go->area.height);
+	}
+
+
+	/* Prepare for framebuffer */
+	glGenTextures(1, &texture);
+	glBindTexture(GL_TEXTURE_2D, texture);
+	glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, cap_buf_image);
+	glBindTexture(GL_TEXTURE_2D, 0);
+
+	glGenFramebuffers(1, &framebuffer);
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+	glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texture, 0);
+	status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
+	if (status != GL_FRAMEBUFFER_COMPLETE) {
+		glDeleteFramebuffers(1, &framebuffer);
+		glDeleteTextures(1, &texture);
+		weston_log("Error! can't make FBO.\n");
+		return;
+	}
+
+	/* Calculate the viewport */
+	glViewport(go->borders[GL_RENDERER_BORDER_LEFT].width,
+			   go->borders[GL_RENDERER_BORDER_BOTTOM].height,
+			   gbm_buf->width,
+			   gbm_buf->height);
+
+	/* Calculate the global GL matrix */
+	go->output_matrix = output->matrix;
+	weston_matrix_translate(&go->output_matrix,
+				-(output->current_mode->width / 2.0),
+				-(output->current_mode->height / 2.0), 0);
+	/* Change y to y_invert */
+	weston_matrix_scale(&go->output_matrix,
+			    2.0 / output->current_mode->width,
+			    2.0 / output->current_mode->height, 1);
+
+	glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+	glEnableVertexAttribArray(0);
+
+	/* Draw all views */
+	wl_list_for_each_reverse(pnode, &output->paint_node_z_order_list,
+				 z_order_link) {
+		if (screen_capture_backend->is_screen_capture_view(pnode->view))
+			continue;
+		draw_paint_node(pnode, &rb->base.damage);
+	}
+
+	draw_output_borders(output, rb->border_damage);
+
+	glDisableVertexAttribArray(0);
+
+	/* Check if FBO rendering is completed. Any efficient way except glFinish? */
+
+	glFinish();
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+	glDeleteFramebuffers(1, &framebuffer);
+	glDeleteTextures(1, &texture);
+
+	go->border_status = BORDER_STATUS_CLEAN;
+}
+
 static int
 gl_renderer_read_pixels(struct weston_output *output,
 			const struct pixel_format_info *format, void *pixels,
@@ -3124,6 +3425,306 @@ gl_renderer_attach_dmabuf(struct weston_surface *surface,
 	return true;
 }
 
+static EGLImageKHR
+import_simple_gbm_buffer(struct gl_renderer *gr,struct gbm_buffer *gbmbuf)
+{
+	EGLImageKHR egl_image;
+	EGLint attribs[50];
+	int atti = 0;
+	int colorspace = 0;
+	ColorMetaData colormeta = {0};
+	int result = -1;
+
+	/* This requires the Mesa commit in
+	 * Mesa 10.3 (08264e5dad4df448e7718e782ad9077902089a07) or
+	 * Mesa 10.2.7 (55d28925e6109a4afd61f109e845a8a51bd17652).
+	 * Otherwise Mesa closes the fd behind our back and re-importing
+	 * will fail.
+	 * https://bugs.freedesktop.org/show_bug.cgi?id=76188
+	 */
+	attribs[atti++] = EGL_WIDTH;
+	attribs[atti++] = gbmbuf->width;
+	attribs[atti++] = EGL_HEIGHT;
+	attribs[atti++] = gbmbuf->height;
+	attribs[atti++] = EGL_LINUX_DRM_FOURCC_EXT;
+	attribs[atti++] = gbmbuf->format;
+	/* XXX: Add modifier here when supported */
+	if (gbmbuf->num_planes > 0) {
+		attribs[atti++] = EGL_DMA_BUF_PLANE0_FD_EXT;
+		attribs[atti++] = gbmbuf->fd;
+		attribs[atti++] = EGL_DMA_BUF_PLANE0_OFFSET_EXT;
+		attribs[atti++] = gbmbuf->offset[0];
+		attribs[atti++] = EGL_DMA_BUF_PLANE0_PITCH_EXT;
+		attribs[atti++] = gbmbuf->stride[0];
+	}
+	if (gbmbuf->num_planes > 1) {
+		attribs[atti++] = EGL_DMA_BUF_PLANE1_FD_EXT;
+		attribs[atti++] = -1;
+		attribs[atti++] = EGL_DMA_BUF_PLANE1_OFFSET_EXT;
+		attribs[atti++] = gbmbuf->offset[1];
+		attribs[atti++] = EGL_DMA_BUF_PLANE1_PITCH_EXT;
+		attribs[atti++] = gbmbuf->stride[1];
+	}
+	if (gbmbuf->num_planes > 2) {
+		attribs[atti++] = EGL_DMA_BUF_PLANE2_FD_EXT;
+		attribs[atti++] = -1;
+		attribs[atti++] = EGL_DMA_BUF_PLANE2_OFFSET_EXT;
+		attribs[atti++] = gbmbuf->offset[2];
+		attribs[atti++] = EGL_DMA_BUF_PLANE2_PITCH_EXT;
+		attribs[atti++] = gbmbuf->stride[2];
+	}
+
+	result = gbm_perform(GBM_PERFORM_GET_METADATA,
+			gbmbuf->bo,
+			GBM_METADATA_GET_COLOR_METADATA,
+			&colormeta);
+	if (result == GBM_ERROR_NONE) {
+		switch (colormeta.matrixCoefficients) {
+		case MatrixCoEff_BT709_5:
+			/* currently don't support the 709 full range yet.*/
+			colorspace = (colormeta.range) ? GBM_METADATA_COLOR_SPACE_ITU_R_709 :
+							GBM_METADATA_COLOR_SPACE_ITU_R_709;
+			break;
+		case MatrixCoEff_BT601_6_525:
+		case MatrixCoEff_BT601_6_625:/* fall back to BT601_6_525 case currently.*/
+			colorspace = (colormeta.range) ? GBM_METADATA_COLOR_SPACE_ITU_R_601_FR :
+							GBM_METADATA_COLOR_SPACE_ITU_R_601;
+			break;
+		case MatrixCoEff_BT2020:
+			colorspace = (colormeta.range) ? GBM_METADATA_COLOR_SPACE_ITU_R_2020_FR :
+							GBM_METADATA_COLOR_SPACE_ITU_R_2020;
+			break;
+		default:
+			break;
+		}
+	} else {
+		result = gbm_perform(GBM_PERFORM_GET_METADATA,
+				gbmbuf->bo,
+				GBM_METADATA_GET_COLOR_SPACE,
+				&colorspace);
+	}
+	if (result == GBM_ERROR_NONE) {
+		switch (colorspace) {
+		case GBM_METADATA_COLOR_SPACE_ITU_R_601:
+			attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+			attribs[atti++] = EGL_ITU_REC601_EXT;
+			attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+			attribs[atti++] = EGL_YUV_NARROW_RANGE_EXT;
+			break;
+		case GBM_METADATA_COLOR_SPACE_ITU_R_601_FR:
+			attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+			attribs[atti++] = EGL_ITU_REC601_EXT;
+			attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+			attribs[atti++] = EGL_YUV_FULL_RANGE_EXT;
+			break;
+		case GBM_METADATA_COLOR_SPACE_ITU_R_709:
+			attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+			attribs[atti++] = EGL_ITU_REC709_EXT;
+			attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+			attribs[atti++] = EGL_YUV_NARROW_RANGE_EXT;
+			break;
+		case GBM_METADATA_COLOR_SPACE_ITU_R_2020:
+			attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+			attribs[atti++] = EGL_ITU_REC2020_EXT;
+			attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+			attribs[atti++] = EGL_YUV_NARROW_RANGE_EXT;
+			break;
+		case GBM_METADATA_COLOR_SPACE_ITU_R_2020_FR:
+			attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+			attribs[atti++] = EGL_ITU_REC2020_EXT;
+			attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+			attribs[atti++] = EGL_YUV_FULL_RANGE_EXT;
+			break;
+		default:
+			weston_log("Unsupport GBM_METADATA_COLOR_SPACE:%d\n", colorspace);
+			break;
+		}
+	}
+
+	attribs[atti++] = EGL_NONE;
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->width=%d", gbmbuf->width);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->height=%d", gbmbuf->height);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->format=%d", gbmbuf->format);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->num_planes=%d", gbmbuf->num_planes);
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->fd=%d\n", gbmbuf->fd);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->offset[0]=%d", gbmbuf->offset[0]);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->stride[0]=%d", gbmbuf->stride[0]);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->offset[1]=%d", gbmbuf->offset[1]);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->stride[1]=%d", gbmbuf->stride[1]);
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->offset[2]=%d", gbmbuf->offset[2]);
+	GBM_PROTOCOL_LOG(LOG_DBG,"gbmbuf->stride[2]=%d", gbmbuf->stride[2]);
+
+	egl_image = gr->create_image(gr->egl_display, EGL_NO_CONTEXT,
+					EGL_LINUX_DMA_BUF_EXT, NULL, attribs);
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"import_simple_gbm_buffer::Image created =%p\n", egl_image);
+
+	return egl_image;
+
+}
+
+static GLenum
+choose_texture_gbm_buf_target(struct gbm_buffer *gbmbuf)
+{
+	if (gbmbuf->num_planes > 1)
+		return GL_TEXTURE_EXTERNAL_OES;
+
+	switch (gbmbuf->format & ~DRM_FORMAT_BIG_ENDIAN) {
+	case DRM_FORMAT_YUYV:
+	case DRM_FORMAT_YVYU:
+	case DRM_FORMAT_UYVY:
+	case DRM_FORMAT_VYUY:
+	case DRM_FORMAT_AYUV:
+		return GL_TEXTURE_EXTERNAL_OES;
+	default:
+		return GL_TEXTURE_2D;
+	}
+}
+
+static void
+gl_renderer_destroy_gbm_buffer(struct gbm_buffer *gbm_buf)
+{
+	struct gl_buffer_state *gb =
+			gbm_buffer_backend->get_user_data(gbm_buf);
+
+	gbm_buffer_backend->set_user_data(gbm_buf, NULL, NULL);
+	destroy_buffer_state(gb);
+}
+
+static struct gl_buffer_state *
+import_gbm_buffer(struct gl_renderer *gr,struct gbm_buffer *gbmbuf)
+{
+	EGLImageKHR egl_image;
+	struct gl_buffer_state *gb;
+	GLenum target;
+
+	gb = zalloc(sizeof(*gb));
+	if (!gb)
+		return NULL;
+
+	gb->gr = gr;
+	pixman_region32_init(&gb->texture_damage);
+	wl_list_init(&gb->destroy_listener.link);
+
+	egl_image = import_simple_gbm_buffer(gr, gbmbuf);
+
+	if (egl_image != EGL_NO_IMAGE_KHR) {
+		gb->num_images = 1;
+		gb->images[0] = egl_image;
+		target = choose_texture_gbm_buf_target(gbmbuf);
+
+		switch (target) {
+		case GL_TEXTURE_2D:
+			gb->shader_variant = SHADER_VARIANT_RGBA;
+			break;
+		default:
+			gb->shader_variant = SHADER_VARIANT_EXTERNAL;
+		}
+
+		ensure_textures(gb, target, gb->num_images);
+	} else {
+		destroy_buffer_state(gb);
+		return NULL;
+	}
+
+	return gb;
+}
+
+static bool
+gl_renderer_import_gbm_buffer(struct weston_compositor *ec, struct gbm_buffer *gbm_buf)
+{
+	struct gl_renderer *gr = get_renderer(ec);
+	struct gl_buffer_state *gb;
+	struct gbm_device * gbm = gr->gbm_hdle;
+	struct gbm_buf_info  buf_info;
+	generic_buf_layout_t buf_lyt;
+	struct gbm_bo *bo;
+	uint32_t j;
+
+	buf_info.fd		  = gbm_buf->fd;
+	buf_info.metadata_fd = gbm_buf->metadata_fd;
+	buf_info.height	  = gbm_buf->height;
+	buf_info.width	   = gbm_buf->width;
+	buf_info.format	  = gbm_buf->format;
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gl_renderer_import_gbm_buffer:Invoked");
+
+	//We will import BO to create an entry into the hash map fo this fd
+	bo = gbm_bo_import(gbm, GBM_BO_IMPORT_GBM_BUF_TYPE, &buf_info, GBM_BO_USE_RENDERING);
+	//save gbm buffer object
+	gbm_buf->bo = bo;
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gl_renderer_import_gbm_buffer:bo created= %p",bo);
+
+	int ret=gbm_perform(GBM_PERFORM_GET_PLANE_INFO, bo, &buf_lyt);
+	if (ret == GBM_ERROR_NONE){
+		gbm_buf->num_planes = buf_lyt.num_planes;
+		for(j = 0; j < buf_lyt.num_planes; j++){
+			gbm_buf->offset[j] = buf_lyt.planes[j].offset;
+			gbm_buf->stride[j] = buf_lyt.planes[j].v_increment;
+		}
+	} else {
+		weston_log("gl_renderer_import_gbm_buffer::GET YUV Info failed\n");
+		return false;
+	}
+
+	//Fill up the remaining fields with default values
+	for (; j < MAX_NUM_PLANES; j++){
+		gbm_buf->offset[j] = 0;
+		gbm_buf->stride[j] = 0;
+	}
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gl_renderer_import_gbm_buffer:Invoke import_gbm_buffer()");
+
+	gb = import_gbm_buffer(gr, gbm_buf);
+	if (!gb)
+		return false;
+
+	gbm_buffer_backend->set_user_data(gbm_buf, gb, gl_renderer_destroy_gbm_buffer);
+
+	GBM_PROTOCOL_LOG(LOG_DBG,"gl_renderer_import_gbm_buffer:Exited");
+	return true;
+}
+
+static bool
+gl_renderer_attach_gbm_buffer(struct weston_surface *surface,
+	          struct weston_buffer *buffer)
+{
+	struct gl_renderer *gr = get_renderer(surface->compositor);
+	struct gl_surface_state *gs = get_surface_state(surface);
+	struct gl_buffer_state *gb;
+	struct gbm_buffer *gbmbuf = buffer->gbmbuf;
+	GLenum target;
+	int i, ret;
+
+	if (!buffer->renderer_private) {
+		gb = gbm_buffer_backend->get_user_data(gbmbuf);
+		assert(gb);
+		gbm_buffer_backend->set_user_data(gbmbuf, NULL, NULL);
+		buffer->renderer_private = gb;
+		gb->destroy_listener.notify = handle_buffer_destroy;
+		wl_signal_add(&buffer->destroy_signal, &gb->destroy_listener);
+	}
+
+	assert(buffer->renderer_private);
+	assert(gbm_buffer_backend->get_user_data(gbmbuf) == NULL);
+	gb = buffer->renderer_private;
+
+	gs->buffer = gb;
+	target = gl_shader_texture_variant_get_target(gb->shader_variant);
+
+	for (i = 0; i < gb->num_images; ++i) {
+		glActiveTexture(GL_TEXTURE0 + i);
+		glBindTexture(target, gb->textures[i]);
+		gr->image_target_texture_2d(target, gb->images[i]);
+	}
+
+	return true;
+}
+
 static const struct weston_drm_format_array *
 gl_renderer_get_supported_formats(struct weston_compositor *ec)
 {
@@ -3214,12 +3815,15 @@ static void
 gl_renderer_attach(struct weston_surface *es, struct weston_buffer *buffer)
 {
 	struct gl_surface_state *gs = get_surface_state(es);
+	struct gbm_buffer *gbmbuf;
+	struct weston_compositor *ec = es->compositor;
 	bool ret = false;
 
 	/* If get_surface_state called gl_renderer_create_surface, it did
 	 * attach the buffer */
-	if (gs->buffer_ref.buffer == buffer)
+	if (gs->buffer_ref.buffer == buffer) {
 		return;
+	}
 
 	/* SHM buffers are a little special in that they are allocated
 	 * per-surface rather than per-buffer, because we keep a shadow
@@ -3247,6 +3851,9 @@ gl_renderer_attach(struct weston_surface *es, struct weston_buffer *buffer)
 	case WESTON_BUFFER_DMABUF:
 		ret = gl_renderer_attach_dmabuf(es, buffer);
 		break;
+	case WESTON_BUFFER_GBMBUF:
+		ret = gl_renderer_attach_gbm_buffer(es, buffer);
+		break;
 	case WESTON_BUFFER_RENDERER_OPAQUE:
 		ret = gl_renderer_attach_egl(es, buffer);
 		break;
@@ -3265,9 +3872,14 @@ gl_renderer_attach(struct weston_surface *es, struct weston_buffer *buffer)
 	}
 
 	weston_buffer_reference(&gs->buffer_ref, buffer,
-				BUFFER_MAY_BE_ACCESSED);
+					BUFFER_MAY_BE_ACCESSED);
+
 	weston_buffer_release_reference(&gs->buffer_release_ref,
 					es->buffer_release_ref.buffer_release);
+
+	/* Create screen capture buffer after creating egl image. FBO path needs it */
+	screen_capture_backend->attach(ec, buffer);
+
 	return;
 
 out:
@@ -3493,6 +4105,24 @@ gl_renderer_create_surface(struct weston_surface *surface)
 						 surface->buffer_ref.buffer,
 						 NULL);
 		}
+
+	/* TODO: To check if it's still needed for early boot path */
+	// } else if (surface->surf_color.is_pended) {
+	// 	/*
+	// 	 * if weston_surface_set_color is called before
+	// 	 * gl renderer is initialized, surface color is stored
+	// 	 * in surf_color and is_pended is set to true. Here,
+	// 	 * gs color is set accordingly.
+	// 	 */
+	// 	gs->color[0] = surface->surf_color.red;
+	// 	gs->color[1] = surface->surf_color.green;
+	// 	gs->color[2] = surface->surf_color.blue;
+	// 	gs->color[3] = surface->surf_color.alpha;
+	// 	//gs->buffer_type = BUFFER_TYPE_SOLID;
+	// 	gs->pitch = 1;
+	// 	gs->height = 1;
+	// 	gs->shader_variant = SHADER_VARIANT_SOLID;
+	// 	surface->surf_color.is_pended = false;
 	}
 
 	return 0;
@@ -3653,6 +4283,11 @@ gl_renderer_create_window_surface(struct gl_renderer *gr,
 {
 	EGLSurface egl_surface = EGL_NO_SURFACE;
 	EGLConfig egl_config;
+	static bool firstCreateWin = true;
+
+	if (firstCreateWin) {
+		weston_place_marker("G - first create window start");
+	}
 
 	egl_config = gl_renderer_get_egl_config(gr, EGL_WINDOW_BIT,
 						formats, formats_count);
@@ -3671,6 +4306,11 @@ gl_renderer_create_window_surface(struct gl_renderer *gr,
 						     egl_config,
 						     window_for_legacy, NULL);
 
+	if (firstCreateWin) {
+		weston_place_marker("G - first create window end");
+		firstCreateWin = false;
+	}
+
 	return egl_surface;
 }
 
@@ -3893,6 +4533,21 @@ gl_renderer_display_create(struct weston_compositor *ec,
 {
 	struct gl_renderer *gr;
 	int ret;
+	static bool firstCreateDisplay = true;
+
+	/* load protocol implementation backends */
+	gbm_buffer_backend = weston_load_module("gbm-buffer-backend.so",
+						"gbm_buffer_backend_c_interface",
+						LIBWESTON_MODULEDIR);
+	assert(gbm_buffer_backend);
+	screen_capture_backend = weston_load_module("screen-capture.so",
+						"screen_capture_c_interface",
+						LIBWESTON_MODULEDIR);
+	assert(screen_capture_backend);
+
+	if (firstCreateDisplay) {
+		weston_place_marker("G - setup egl start");
+	}
 
 	gr = zalloc(sizeof *gr);
 	if (gr == NULL)
@@ -3902,6 +4557,9 @@ gl_renderer_display_create(struct weston_compositor *ec,
 	wl_list_init(&gr->shader_list);
 	gr->platform = options->egl_platform;
 
+	//gbm device handle
+	gr->gbm_hdle =(struct gbm_device *)options->egl_native_display;
+
 	gr->renderer_scope = weston_compositor_add_log_scope(ec, "gl-renderer",
 		"GL-renderer verbose messages\n", NULL, NULL, gr);
 	if (!gr->renderer_scope)
@@ -3916,6 +4574,7 @@ gl_renderer_display_create(struct weston_compositor *ec,
 
 	gr->base.read_pixels = gl_renderer_read_pixels;
 	gr->base.repaint_output = gl_renderer_repaint_output;
+	gr->base.capture_screen = gl_renderer_capture_screen;
 	gr->base.resize_output = gl_renderer_resize_output;
 	gr->base.flush_damage = gl_renderer_flush_damage;
 	gr->base.attach = gl_renderer_attach;
@@ -3956,14 +4615,21 @@ gl_renderer_display_create(struct weston_compositor *ec,
 		}
 	}
 
+	if (firstCreateDisplay) {
+		weston_place_marker("G - setup egl end");
+		firstCreateDisplay = false;
+	}
+
 	ec->capabilities |= WESTON_CAP_ROTATION_ANY;
 	ec->capabilities |= WESTON_CAP_CAPTURE_YFLIP;
 	ec->capabilities |= WESTON_CAP_VIEW_CLIP_MASK;
 	if (gr->has_native_fence_sync && gr->has_wait_sync)
 		ec->capabilities |= WESTON_CAP_EXPLICIT_SYNC;
 
+	wl_list_init(&gr->gbmbuf_images);
 	if (gr->has_dmabuf_import) {
 		gr->base.import_dmabuf = gl_renderer_import_dmabuf;
+		gr->base.import_gbm_buffer = gl_renderer_import_gbm_buffer;
 		gr->base.get_supported_formats = gl_renderer_get_supported_formats;
 		ret = populate_supported_formats(ec, &gr->supported_formats);
 		if (ret < 0)
diff --git a/libweston/renderer-gl/meson.build b/libweston/renderer-gl/meson.build
index 4d9f64d..a3d0f8f 100644
--- a/libweston/renderer-gl/meson.build
+++ b/libweston/renderer-gl/meson.build
@@ -34,7 +34,9 @@ deps_renderer_gl = [
 	dep_pixman,
 	dep_libweston_private,
 	dep_libdrm_headers,
-	dep_vertex_clipping
+	dep_libcutils,
+	dep_vertex_clipping,
+	dep_gbm
 ]
 
 foreach name : [ 'egl', 'glesv2' ]
@@ -55,3 +57,8 @@ plugin_gl = shared_library(
 	install_dir: dir_module_libweston
 )
 env_modmap += 'gl-renderer.so=@0@;'.format(plugin_gl.full_path())
+
+install_headers(
+	'gl-renderer.h',
+	subdir: dir_include_libweston_private_install
+)
\ No newline at end of file
diff --git a/meson.build b/meson.build
index 2383273..08ea58b 100644
--- a/meson.build
+++ b/meson.build
@@ -32,6 +32,7 @@ dir_bin = dir_prefix / get_option('bindir')
 dir_data = dir_prefix / get_option('datadir')
 dir_include_libweston = 'libweston-@0@'.format(libweston_major)
 dir_include_libweston_install = dir_include_libweston / 'libweston'
+dir_include_libweston_private_install = join_paths(dir_include_libweston, 'libweston-private')
 dir_lib = dir_prefix /get_option('libdir')
 dir_libexec = dir_prefix / get_option('libexecdir')
 dir_module_weston = dir_lib / 'weston'
@@ -42,6 +43,8 @@ dir_man = dir_prefix / get_option('mandir')
 dir_protocol_libweston = 'libweston-@0@'.format(libweston_major) / 'protocols'
 dir_sysconf = dir_prefix / get_option('sysconfdir')
 dir_gitlab_ci = meson.project_source_root() / '.gitlab-ci'
+dir_weston_shared = 'weston-shared'
+dir_shared = 'shared'
 
 public_inc = include_directories('include')
 common_inc = [ include_directories('.'), public_inc ]
@@ -143,6 +146,8 @@ dep_libdl = cc.find_library('dl')
 dep_libdrm = dependency('libdrm', version: '>= 2.4.108')
 dep_libdrm_headers = dep_libdrm.partial_dependency(compile_args: true)
 dep_threads = dependency('threads')
+dep_libcutils = cc.find_library('cutils')
+dep_gbm = dependency('gbm')
 
 dep_lcms2 = dependency('lcms2', version: '>= 2.9', required: false)
 
diff --git a/shared/meson.build b/shared/meson.build
index 86d46a0..2d3e168 100644
--- a/shared/meson.build
+++ b/shared/meson.build
@@ -106,3 +106,11 @@ dep_matrix_c = declare_dependency(
 	include_directories: public_inc,
 	dependencies: dep_libm
 )
+
+install_headers(
+	'helpers.h',
+	'timespec-util.h',
+	'weston-egl-ext.h',
+	'string-helpers.h',
+	subdir: dir_shared
+)
\ No newline at end of file
